pkg/obfuscate/appsec.go

```diff
@@ -0,0 +1,538 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package obfuscate
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+// ObfuscateAppSec obfuscates the given appsec tag value in order to remove
+// sensitive values from the appsec security events.
+func (o *Obfuscator) ObfuscateAppSec(val string) string {
+	keyRE := o.opts.AppSec.ParameterKeyRegexp
+	valueRE := o.opts.AppSec.ParameterValueRegexp
+	if keyRE == nil && valueRE == nil {
+		return val
+	}
+	ob := appsecEventsObfuscator{
+		keyRE:   keyRE,
+		valueRE: valueRE,
+	}
+	output, err := ob.obfuscate(val)
+	if err != nil {
+		o.log.Errorf("unexpected error while obfuscating the appsec events: %v", err)
+		return val
+	}
+	return output
+}
+
+type appsecEventsObfuscator struct {
+	keyRE, valueRE *regexp.Regexp
+}
+
+type unexpectedScannerOpError int
+
+func (err unexpectedScannerOpError) Error() string {
+	return fmt.Sprintf("unexpected json scanner operation %d", err)
+}
+
+// inputDiff holds the list of diffs, built by the obfuscator, and to ultimately
+// apply to the input.
+type inputDiff []struct {
+	from, to int
+	value    string
+}
+
+// Add a new entry to the list. Note that due to how the apply() method works
+// by creating the output string by walking this list, the list entries must be
+// sorted in ascending "input offset" order.
+func (d *inputDiff) add(from, to int, value string) {
+	elt := struct {
+		from  int
+		to    int
+		value string
+	}{
+		from:  from,
+		to:    to,
+		value: value,
+	}
+	diff := *d
+	l := len(diff)
+	i := sort.Search(l, func(i int) bool {
+		return (*d)[i].to > from
+	})
+	if i == l {
+		*d = append(diff, elt)
+		return
+	}
+	*d = append(diff[:i+1], diff[i:]...)
+	(*d)[i] = elt
+}
+
+// Apply the diff to the input if any.
+func (d inputDiff) apply(input string) string {
+	from := 0
+	var output strings.Builder
+	for _, diff := range d {
+		output.WriteString(input[from:diff.from])
+		from = diff.to + 1
+		output.WriteString(diff.value)
+	}
+	output.WriteString(input[from:])
+	return output.String()
+}
+
+// Append the given diff to the current one, shifted by the given offset.
```

Comment:
Please check your method names vs. documentation, they are different. Also, I think you were better off the first part. In an unexported structure, it's sometimes nicer to differentiate between internal and non-internal methods by capitalising the non-internal ones. No strong opinions here. Just a thought.

Also, this might be better called `merge`. "append" generally is at the end. In your case, things don't go at the end, they get sorted.

Reply:
Not sure what you mean: do you want those comments to start with a lower-case letter to follow the same coding-style as public identifiers?